### PR TITLE
451001 - Natalia Naumova

### PR DESCRIPTION
**Pull request by NataliaNaumova**

### CODE REVIEW

File: DTOGenerator/ClassDescription.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+namespace DTOGenerator
+{
+    public class ClassDescription
+    {
+        public string ClassName { get; private set; }
+        public List<PropertyDescription> PropertyDescriptions { get; private set; }
+
+        public ClassDescription(string className)
+        {
+            this.ClassName = className;
```
Comment (RamanBut-Husaim) [inline DTOGenerator/ClassDescription.cs]: @NataliaNaumova Inconsistent `this` usage.

File: DTOGenerator/DTOGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                            allThreadsAreBusy = false;
+                        }
+                    }
+                }
+
+                ThreadPool.QueueUserWorkItem(delegate (object state)
+                {
+                    try
+                    {
+                        workingThreadsCount++;
```
Comment (RamanBut-Husaim) [inline DTOGenerator/DTOGenerator.cs]: @NataliaNaumova Increment variable in multiple threads in a non-thread safe manner.

File: DTOGenerator/DTOGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            List<DTODescription> classUnits = new List<DTODescription>();
+            CountdownEvent countDownEvent = new CountdownEvent(classesList.Count);
+            Boolean allThreadsAreBusy = false;
+
+            foreach (ClassDescription classDescription in classesList)
+            {
+                DTODescription dtoDescription = new DTODescription(classDescription.ClassName);
+                classUnits.Add(dtoDescription);
+
+                if (workingThreadsCount == threadPoolLimit)
```
Comment (RamanBut-Husaim) [inline DTOGenerator/DTOGenerator.cs]: @NataliaNaumova Your main thread in running in an potentially infinite loop while the the thread pool threads are working? And if the code generation approach is extremely long running? Don't you think that this is a tremendous waste of resources?

File: DTOGenerator/SimpleLockThreadPool.cs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Threading;
+using System.Security.Permissions;
+
+namespace DTOGenerator
+{
+    public class SimpleLockThreadPool : IThreadPool
```
Comment (RamanBut-Husaim) [inline DTOGenerator/SimpleLockThreadPool.cs]: @NataliaNaumova Be ready to answer how this stuff is working.

Comment (RamanBut-Husaim) [inline DTOGenerator/SimpleLockThreadPool.cs]: @NataliaNaumova Are you using it in your lab?

File: DTOGenerator/DTOGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        {
+            this.dtoNamespace = dtoNamespace;
+            this.threadPoolLimit = threadPoolLimit;
+            workingThreadsCount = 0;
+        }
+
+        public List<DTODescription> GenerateCode(List<ClassDescription> classesList)
+        {
+            List<DTODescription> classUnits = new List<DTODescription>();
+            CountdownEvent countDownEvent = new CountdownEvent(classesList.Count);
```
Comment (RamanBut-Husaim) [inline DTOGenerator/DTOGenerator.cs]: @NataliaNaumova This is a disposable object so you should clean the resource properly.

File: DTOGenerator/SimpleLockThreadPool.cs (hunk 4 above)
Comment (NataliaNaumova) [inline DTOGenerator/SimpleLockThreadPool.cs]: No, sorry, I'll delete it.

**Pull request merged**